### ISSUE TITLE
Option to add CORS headers to static assets

### ISSUE THREAD

**Issue by hubermat** [nuxt/framework#4009]
### Discussed in https://github.com/nuxt/framework/discussions/4003

<div type='discussions-op-text'>

<sup>Originally posted by **hubermat** March 31, 2022</sup>
Hello,

our Nuxt-based project serves, besides other API routes, static files (i.e. files from `./public`). These static files will also be accessed from outside our web application, but browsers apply CORS policies. Up to now I could not find a way to add `access-control-allow-origin` headers to the static files that are served by the Nuxt server.

I have created a pull request, that adds a new config option in nuxt.config.js:

```js
render: {
  static: {
    /**
     * Whether to add headers to allow access from any origin
     *
     * If set to true, all responses of static files from the `public/` directory will contain the following headers:
     * Access-Control-Allow-Origin *
     * Access-Control-Allow-Methods GET
     * Access-Control-Allow-Headers Origin, X-Requested-With, Content-Type, Accept
     * X-Frame-Options ALLOWALL
     */
    allowFromAnyOrigin: false
  },
}
```

If allowFromAnyOrigin is set to true, for development mode, a function setHeaders will be passed to serve-static to set these headers. For production mode, the public assets will be flagged and the headers will be set in the static middleware.
</div>

**Comment by chengzhuo5** [comment]
It`s useful for me:
```ts
import { defineCorsEventHandler } from '@nozomuikuta/h3-cors';
export default defineNuxtConfig({
  builder: 'webpack',
  devServerHandlers: [
    {
      route: '/',
      handler: defineCorsEventHandler({}),
    },
  ],
  // With vite
  vite: {
    server: {
      headers: {
        'Access-Control-Allow-Origin': '*',
      }
    }
  },
  // or webpack
  webpack: {
    devMiddleware: {
      headers: {
        'Access-Control-Allow-Origin': '*',
      },
    },
  },
});
```

**Comment by pi0** [comment]
You can try routeRules to set CORS for `/_nuxt` assets: https://v3.nuxtjs.org/guide/concepts/rendering#route-rules. For your own, create a sub directory inside public dir like `public/static` and set `/static: { cors: true }` rule.

**Comment by chengzhuo5** [comment]
It`s invalid with webpack builder.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "nuxt/framework"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年10月21日(星期五) 下午4:11
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [nuxt/framework] Option to add CORS headers to static assets (Issue #4009)





 
You can try routeRules to set CORS for /_nuxt assets: https://v3.nuxtjs.org/guide/concepts/rendering#route-rules. For your own, create a sub directory inside public dir like public/static and set /static: { cors: true } rule.

**Comment by chengzhuo5** [comment]
```ts
export default defineNuxtConfig({
  builder: 'webpack',
  // Only valid on preview
  routeRules: {
    '/**': { headers: { 'Access-Control-Allow-Origin': '*' } },
  },
  // Only valid on serve
  webpack: {
    devMiddleware: {
      headers: {
        'Access-Control-Allow-Origin': '*',
      },
    },
  },
});
```

**Comment by rezaseyedi91** [comment]
> You can try routeRules to set CORS for `/_nuxt` assets: https://v3.nuxtjs.org/guide/concepts/rendering#route-rules. For your own, create a sub directory inside public dir like `public/static` and set `/static: { cors: true }` rule.

@pi0 
How can I use this solution when I'm trying to send requests with credentials and get:
`Access to XMLHttpRequest at 'http://localhost:3000/api/v1/admin/auth/login' from origin 'http://localhost:8080' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: The value of the 'Access-Control-Allow-Origin' header in the response must not be the wildcard '*' when the request's credentials mode is 'include'. The credentials mode of requests initiated by the XMLHttpRequest is controlled by the withCredentials attribute.`

I have set nuxt.config/routeRules to `    routeRules: {
      '/**': { headers: { 'Access-Control-Allow-Origin': 'http://localhost:8080' } },
    },`
and didn't work out. 
I appreciate any guidance on this.

**Comment by P4sca1** [comment]
Nuxt RC.12 automatically adds `Access-Control-Origin: *` header. For security reasons, and due to the issue mentioned above, I think this is a bad idea. @danielroe

**Comment by danielroe** [comment]
You should be able to override with a custom route rule. If not, please do open a new issue.

You can also not use the built-in `cors` rule and simply set your own headers via the `headers` route rule.

**Comment by P4sca1** [comment]
@rezaseyedi91 This is likely an issue that only occurs in development when using Vite.
You can fix this by explicitly setting your origin in `nuxt.config.ts`:
```ts
	vite: {
		server: {
			// Nuxt launches Vite in middleware mode (port 3000), as well as the HMR server (port 24678).
			// The browser will reject requests from localhost:3000 to the HMR server, due to CORS.
			// Because of this, Vite by default adds "Access-Control-Allow-Origin": "*" header.
			// This is problematic, because fetch calls with credentials: 'include' will get rejected when
			// the origin is not explicitly allowed. Because of this, we need to manually define the origin.
			cors: {
				origin: url,
			},
		},
	},
```
